File: Java/FutureHolderSynchronizer/src/main/java/FutureHolder.java
/**
 * Created by Ruben Gomes on 20/05/2015.
 */
public class FutureHolder<T> implements IFutureHolder<T> {

    private T value ;
   private final Object lock = new Object();

    @Override
    public void setValue(T value) throws IllegalStateException{
        synchronized (lock){
            if(isValueAvailable()){ // if value already inserted throw IllegalStateException
                throw new IllegalStateException("The value already inserted !");
            }else{ // if not , set value and notify waiting threads
                this.value = value;
                lock.notifyAll();
            }
        }
    }

    @Override
    public T getValue(long timeout) throws InterruptedException {
        synchronized (lock){
            if(isValueAvailable()) return value;// if value is available return immediately
            if(timeout == 0) return null;
            long lastTime = System.currentTimeMillis();
            do{
                try{
                    lock.wait(timeout);
                } catch (InterruptedException e) {
                    if(isValueAvailable()){ // if interrupt and value is available interrupt thread and return value
                        Thread.currentThread().interrupt();
                        return value;
                    }
                   throw e;
                }
                if(isValueAvailable()){
                    return value;
                }
                if(timeout > 0) {
                    long currTime = System.currentTimeMillis();
                    long deltaTime = currTime - lastTime;
                    timeout = (deltaTime >= timeout) ? 0 : timeout - deltaTime;
                }
            }while(timeout > 0);
            return null;
        }

    }

    @Override
    public boolean isValueAvailable() {
        return this.value != null;
    }
}

File: C#/FutureHolderSynchronizer/FutureHolderSynchronizer/FutureHolder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace FutureHolderSynchronizer
{
    class FutureHolder<T> : IFutureHolder<T>
    {
        private T value;
        private readonly Object _lock = new Object();
        public void SetValue(T value)
        {
            lock (_lock)
            {
                if (IsValueAvailable())
                {
                    throw new InvalidOperationException();
                }
                else
                {
                    this.value = value;
                    Monitor.PulseAll(_lock);
                }
            }
        }

       

        public T GetValue(long timeout)
        {
            lock (_lock)
            {
                if(IsValueAvailable()) return value;
                if (timeout == 0) return default(T);
                int lastTime = (timeout != Timeout.Infinite) ? Environment.TickCount : 0;
                do
                {
                    try
                    {
                        Monitor.Wait(_lock, (int)timeout);
                    }
                    catch (ThreadInterruptedException)
                    {
                        if (IsValueAvailable())
                        {
                            Thread.CurrentThread.Interrupt();
                            return value;
                        }
                        throw;
                    }
                    if (IsValueAvailable())
                    {
                        return value;
                    }
                    if (SyncUtils.AdjustTimeout(ref lastTime, ref timeout) == 0) 
                        return default(T);

                } while (true);
            }
        }

        public bool IsValueAvailable()
        {
            return this.value != null;
        }
    }
}
